futures-util/src/try_stream/try_zip.rs

```diff
@@ -0,0 +1,122 @@
+use crate::stream::{StreamExt, Fuse};
+use crate::try_stream::{TryStreamExt, IntoStream};
+use core::pin::Pin;
+use futures_core::stream::{FusedStream, TryStream};
+use futures_core::task::{Context, Poll};
+use pin_utils::{unsafe_pinned, unsafe_unpinned};
+
+/// Stream for the [`try_zip`](super::TryStreamExt::try_zip) method.
+#[derive(Debug)]
+#[must_use = "streams do nothing unless polled"]
+pub struct TryZip<St1: TryStream, St2: TryStream> {
+    stream1: Fuse<IntoStream<St1>>,
+    stream2: Fuse<IntoStream<St2>>,
+    queued1: Option<Result<St1::Ok, St1::Error>>,
+    queued2: Option<Result<St2::Ok, St2::Error>>,
+}
+
+impl<St1: TryStream + Unpin, St2: TryStream + Unpin> Unpin for TryZip<St1, St2> {}
```

Comment:
This should technically be

```rust
impl<St1: TryStream, St2: TryStream> Unpin for TryZip<St1, St2>
where
    Fuse<IntoStream<St1>>: Unpin,
    Fuse<IntoStream<St2>>: Unpin,
{}
```

In case either `Fuse` or `IntoStream` don’t have the equivalent `Unpin` implementations.

Reply:
Makes sense, but it's unfortunate that the internal implementation details need to leak out like that.

Should the `impl Unpin for Zip` be updated to use `Fuse` as well? Is this a broader issue?

Reply:
> it's unfortunate that the internal implementation details need to leak out like that.

Yeah, that was my feeling as well as I wrote the comment. I rationalized it as being the same as the auto-trait's structural rules, if `TryZip` were to rely on the implicit `Unpin` you would still get error messages saying "`Fuse` does not implement `Unpin`".

> Is this a broader issue?

Probably, I'll try to remember to open a ticket about it. (This is one of the reasons I prefer [`pin-project`](https://docs.rs/pin-project/) for projection, it can derive the `Unpin` implementation for you so that you don't have to keep it consistent with the current field definitions).

Reply:
I was just looking into this a little more as I encountered the situation in another library where the `Unpin` definition depends on a crate internal type. It looks like rustdoc has special handling for the automatic implementations that looks through the field types and shows the bounds that result from them instead of revealing the internal types, it would be nice if there were some way to request that to happen for manual implementations as well.

For example

```rust
struct Foo<T>(T);

pub struct Bar<T>(Foo<T>);

unsafe impl<T: Send + Iterator> Send for Foo<T> {}
```

shows in the docs

```rust
impl<T> Send for Bar<T>
where
    T: Iterator + Send, 
```